docker/pusher.go

```diff
@@ -2,67 +2,142 @@ package docker
 
 import (
 	"bytes"
-	"errors"
+	"context"
+	"encoding/base64"
+	"encoding/json"
 	"fmt"
+
 	bosonFunc "github.com/boson-project/func"
+
 	"io"
 	"os"
-	"os/exec"
 	"regexp"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/pkg/errors"
 )
 
+type Opt func(*Pusher) error
+
+type Credentials struct {
+	Username string
+	Password string
+}
+
+type CredentialsProvider func(ctx context.Context, registry string) (Credentials, error)
+
 // Pusher of images from local to remote registry.
 type Pusher struct {
 	// Verbose logging.
-	Verbose bool
+	Verbose 			bool
+	credentialsProvider CredentialsProvider
+}
+
+func WithCredentialsProvider(cp CredentialsProvider) Opt {
+	return func(p *Pusher) error {
+		p.credentialsProvider = cp
+		return nil
+	}
+}
+
+func EmptyCredentialsProvider(ctx context.Context, registry string) (Credentials, error) {
+	return Credentials{}, nil
 }
 
 // NewPusher creates an instance of a docker-based image pusher.
-func NewPusher() *Pusher {
-	return &Pusher{}
+func NewPusher(opts ...Opt) (*Pusher, error) {
+	result := &Pusher{
+		Verbose: false,
+		credentialsProvider: EmptyCredentialsProvider,
+	}
+	for _, opt := range opts {
+		err := opt(result)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return result, nil
 }
 
 // Push the image of the Function.
-func (n *Pusher) Push(f bosonFunc.Function) (digest string, err error) {
-	// Check for the docker binary explicitly so that we can return
-	// an extra-friendly error message.
-	_, err = exec.LookPath("docker")
-	if err != nil {
-		err = errors.New("please install 'docker'")
-		return
-	}
+func (n *Pusher) Push(ctx context.Context, f bosonFunc.Function) (digest string, err error) {
 
 	if f.Image == "" {
 		return "", errors.New("Function has no associated image.  Has it been built?")
 	}
 
-	// set up the command, specifying a sanitized project name and connecting
-	// standard output and error.
-	cmd := exec.Command("docker", "push", f.Image)
+	var registry string
+	parts := strings.Split(f.Image, "/")
+	switch len(parts) {
+	case 2:
+		registry = "docker.io"
```

Comment:
Really nice to see we're preserving the docker.io default, at least for now, because it's the one which doesn't require manual intervention to expose the auto-created repository for a new Function.

One suggestion, could we move this "magic string" up top to be `const DefaultRegistry = "docker.io"`?  That helps when skimming later